### PR TITLE
Issue #230

### PR DESCRIPTION
**Pull request by gregw**
Implemented #230
Factored out VmRuntimeInterceptor
Use beta settings to configure to clear the request ticket: never, after commit, after close

@ludoch is configuration via the beta settings in appengine-web.xml a good way to support this?

### CODE REVIEW

**Comment by gregw** [comment]
I signed it!

**Comment by ludoch** [comment]
1/ the beta settings were not meant to configure that type of behavior.
2/ based on Arie's finding talking to engineers for the gae new stack, the request ticket will disappear in a very near future, and only the global ticket will be supported.   I am not sure we want to continue trying to optimize for the 2 targets... 
Google hangout soon to chat more?

File: jetty9-compat-base/src/main/java/com/google/apphosting/vmruntime/jetty9/VmRuntimeWebAppContext.java [hunk 1]
```diff
@@ -361,23 +366,7 @@ public RequestContext getRequestContext(Request baseRequest) {
           VmRuntimeUtils.ONE_DAY_IN_MILLIS, defaultEnvironment);
       final HttpOutput httpOutput = request.getResponse().getHttpOutput();
       final HttpOutput.Interceptor nextOutput = httpOutput.getInterceptor();
-      httpOutput.setInterceptor(new Interceptor() {
-        @Override
-        public void write(ByteBuffer content, boolean complete, Callback callback) {
-          requestSpecificEnvironment.clearTicket(); // Request ticket not valid after commit
```
Comment (aozarov) [inline jetty9-compat-base/src/main/java/com/google/apphosting/vmruntime/jetty9/VmRuntimeWebAppContext.java]: Help me understand. So, in the current code (where we don't buffer/delay the applications writes) this is going to clear the ticket after the first write to the servlet's outputstream?

File: jetty9-compat-base/src/main/java/com/google/apphosting/vmruntime/jetty9/VmRuntimeInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  @Override
+  public void write(ByteBuffer content, boolean last, Callback callback) {
+    if (clearTicket == ClearTicket.ON_COMMIT)
+      environment.clearTicket();
+
+    next.write(content, last, callback);
+
+    if (clearTicket == ClearTicket.ON_LAST && last)
```
Comment (aozarov) [inline jetty9-compat-base/src/main/java/com/google/apphosting/vmruntime/jetty9/VmRuntimeInterceptor.java]: for synchronous calls last is set to `true` once the http request completes?

What happens for asynchronous calls, do they use the same set of interceptors and if so, `last` will be set to `false` only once `AsyncContext.complete` is called?

File: jetty9-compat-base/src/main/java/com/google/apphosting/vmruntime/jetty9/VmRuntimeInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+  public VmRuntimeInterceptor(VmApiProxyEnvironment environment, HttpOutput.Interceptor next) {
+    this.environment = environment;
+    this.next = next;
+    this.clearTicket = dftClearTicket;
+  }
+
+  @Override
+  public void write(ByteBuffer content, boolean last, Callback callback) {
+    if (clearTicket == ClearTicket.ON_COMMIT)
```
Comment (aozarov) [inline jetty9-compat-base/src/main/java/com/google/apphosting/vmruntime/jetty9/VmRuntimeInterceptor.java]: any reason to clear it before calling next?

**Comment by gregw** [comment]
replaced by #237

**Pull request closed without merging**